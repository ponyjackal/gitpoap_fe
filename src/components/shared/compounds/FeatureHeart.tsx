import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Loader } from '@mantine/core';
import { FaHeart } from 'react-icons/fa';
import { ExtraHover, ExtraPressed, TextAccent, TextLight } from '../../../colors';
import { useFeaturedPOAPs, useFeaturedPOAPsDispatch } from '../../profile/FeaturedPOAPsContext';

type Props = {
  className?: string;
  poapTokenId: string;
};

const HeartWrapper = styled.div<{ isFeatured: boolean }>`
  > svg {
    fill: ${({ isFeatured }) => (isFeatured ? TextAccent : TextLight)};
  }
`;

const FeatureHeartStyled = styled(FaHeart)`
  color: ${TextLight};
  height: ${rem(18)};
  width: ${rem(18)};
  transition: 150ms fill ease;
  fill: ${TextLight};

  &:hover {
    fill: ${ExtraHover};
    cursor: pointer;
  }
  &:active {
    fill: ${ExtraPressed};
  }
`;

export const FeatureHeart = ({ className, poapTokenId }: Props) => {
  const {
    featuredPOAPsState: { featuredPOAPTokenIDs },
    showHearts,
    loadingIds,
  } = useFeaturedPOAPs();
  const { addFeaturedPOAP, removeFeaturedPOAP } = useFeaturedPOAPsDispatch();
  const isFeatured: boolean = !!featuredPOAPTokenIDs && !!featuredPOAPTokenIDs[poapTokenId];
  const isFeaturedLoading = !!loadingIds && !!loadingIds[poapTokenId];

  if (!showHearts) {
    return null;
  }

  return (
    <HeartWrapper isFeatured={isFeatured}>
      {isFeaturedLoading ? (
        <Loader className={className} size={18} />
      ) : (
        <FeatureHeartStyled
          className={className}
          onClick={() => {
            if (isFeatured) {
              removeFeaturedPOAP(poapTokenId);
            } else {
              addFeaturedPOAP(poapTokenId);
            }
          }}
        />
      )}
    </HeartWrapper>
  );
};
