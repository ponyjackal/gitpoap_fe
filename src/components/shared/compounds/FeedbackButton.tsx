import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Popover, Box } from '@mantine/core';
import { useDisclosure } from '@mantine/hooks';
import { GitPOAPLogoNoText } from '../elements/icons';
import { BackgroundPanel, BackgroundPanel2, ExtraHover, ExtraPressed } from '../../../colors';
import { Text } from '../elements';

const StyledLogo = styled(GitPOAPLogoNoText)`
  height: ${rem(45)};
  width: ${rem(45)};
  transition: transform 200ms ease-in-out;
  path {
    transition: fill 200ms ease-in-out;
  }
`;

const CircleButton = styled.a`
  display: flex;
  align-items: center;
  justify-content: center;
  height: ${rem(70)};
  width: ${rem(70)};
  background-color: ${BackgroundPanel};
  transition: background-color 200ms ease-in-out, transform 200ms ease-in-out;
  border-radius: 50%;
  border: ${rem(3)} solid ${BackgroundPanel2};
  &:hover {
    background-color: ${BackgroundPanel2};
    transform: scale(1.1);
    cursor: pointer;
    ${StyledLogo} {
      transform: scale(1.1);
      path {
        fill: ${ExtraHover};
      }
    }
  }
  &:active {
    background-color: ${BackgroundPanel};
    ${StyledLogo} {
      color: ${ExtraHover};
      path {
        fill: ${ExtraPressed};
      }
    }
  }
`;

const PopoverContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

type Props = {
  className?: string;
  href: string;
};

export const FeedbackButton = ({ className, href }: Props) => {
  const [isOpen, { open, close }] = useDisclosure(false);
  return (
    <Popover
      opened={isOpen}
      onClose={close}
      position="top"
      withArrow
      trapFocus={false}
      closeOnEscape={false}
      transition="fade"
      transitionDuration={200}
      radius="lg"
      withinPortal
      zIndex={100}
      styles={{
        dropdown: {
          backgroundColor: BackgroundPanel,
        },
      }}
    >
      <Popover.Target>
        <Box onMouseEnter={open} onMouseLeave={close} className={className} style={{ zIndex: 100 }}>
          <CircleButton href={href} target="_blank" rel="noopener noreferrer">
            <StyledLogo />
          </CircleButton>
        </Box>
      </Popover.Target>
      <Popover.Dropdown>
        <PopoverContainer>
          <Text>{'Give Feedback'}</Text>
        </PopoverContainer>
      </Popover.Dropdown>
    </Popover>
  );
};
