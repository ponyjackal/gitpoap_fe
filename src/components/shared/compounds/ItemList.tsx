import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Header } from '../elements/Header';
import { Button } from '../elements/Button';
import { FaPlus } from 'react-icons/fa';
import { Select } from '../elements/Select';
import { Text } from '../elements/Text';
import { TextGray } from '../../../colors';
import { Input } from '../elements/Input';

type Props = {
  title?: string;
  className?: string;
  children?: React.ReactNode;
  selectOptions: SelectOption[];
  selectValue: string;
  onSelectChange: (value: string) => void;
  isLoading: boolean;
  hasShowMoreButton: boolean;
  showMoreOnClick: () => void;
  searchInputPlaceholder?: string;
  searchInputValue?: string;
  onSearchInputChange?: React.ChangeEventHandler<HTMLInputElement>;
};

export type SelectOption<T = string> = { value: T; label: string };

const Container = styled.div`
  display: inline-flex;
  flex-direction: column;
  width: 100%;
`;

const Heading = styled.div`
  display: inline-flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

const ListTitle = styled(Header)`
  font-size: ${rem(30)};
  line-height: ${rem(42)};
`;

const Sorting = styled.div`
  display: inline-flex;
  flex-direction: row;
`;

const ShowMore = styled(Button)`
  align-self: center;
`;

const SortBy = styled(Text)`
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: ${rem(12)};
  line-height: ${rem(18)};
  letter-spacing: ${rem(2)};
  text-transform: uppercase;
  color: ${TextGray};
  margin-right: ${rem(10)};
`;

const SearchInput = styled(Input)`
  margin-right: ${rem(50)};
`;

export const ItemList = ({
  children,
  className,
  title,
  selectOptions,
  selectValue,
  onSelectChange,
  isLoading,
  hasShowMoreButton,
  showMoreOnClick,
  searchInputPlaceholder,
  searchInputValue,
  onSearchInputChange,
}: Props) => {
  return (
    <Container className={className}>
      <Heading>
        <ListTitle>{title}</ListTitle>
        <Sorting>
          {searchInputValue !== undefined && onSearchInputChange && (
            <SearchInput
              placeholder={searchInputPlaceholder}
              value={searchInputValue}
              onChange={onSearchInputChange}
            />
          )}
          <SortBy>{'Sort By: '}</SortBy>
          <Select data={selectOptions} value={selectValue} onChange={onSelectChange} />
        </Sorting>
      </Heading>
      {children}
      {hasShowMoreButton && (
        <ShowMore
          onClick={showMoreOnClick}
          leftIcon={<FaPlus />}
          variant="outline"
          loading={isLoading}
        >
          {'Show more'}
        </ShowMore>
      )}
    </Container>
  );
};
