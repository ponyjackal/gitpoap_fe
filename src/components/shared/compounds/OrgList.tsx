import styled from 'styled-components';
import { rem } from 'polished';
import { RepoList } from './RepoList';

export const OrgList = styled(RepoList)`
  grid-template-columns: repeat(auto-fill, ${rem(260)});

  @media (max-width: ${rem(800)}) {
    grid-template-columns: unset;
  }
`;
