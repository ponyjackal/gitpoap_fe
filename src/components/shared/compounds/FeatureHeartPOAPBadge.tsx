import React from 'react';
import { Loader } from '../elements/Loader';
import { useFeaturedPOAPsDispatch } from '../../profile/FeaturedPOAPsContext';
import { HeartWrapper, FeatureHeartStyled } from './FeatureHeart';

type Props = {
  className?: string;
  poapTokenId: string;
  isFeatured: boolean;
  isFeaturedLoading: boolean;
};

export const FeaturedHeartPOAPBadge = ({
  className,
  poapTokenId,
  isFeatured,
  isFeaturedLoading,
}: Props) => {
  const { addFeaturedPOAP, removeFeaturedPOAP } = useFeaturedPOAPsDispatch();
  return (
    <HeartWrapper isFeatured={isFeatured}>
      {isFeaturedLoading ? (
        <Loader className={className} size={18} />
      ) : (
        <FeatureHeartStyled
          className={className}
          onClick={() => {
            if (isFeatured) {
              removeFeaturedPOAP(poapTokenId);
            } else {
              addFeaturedPOAP(poapTokenId);
            }
          }}
        />
      )}
    </HeartWrapper>
  );
};
