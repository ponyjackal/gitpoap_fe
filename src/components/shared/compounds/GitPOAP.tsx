import { Link } from '../../Link';
import React from 'react';
import styled, { css } from 'styled-components';
import { rem } from 'polished';
import { GitPOAPBadge } from '../elements/GitPOAPBadge';
import { Title } from '../elements/Title';
import { TextLight } from '../../../colors';
import { FeatureHeart } from './FeatureHeart';

type Props = {
  gitPOAPId: number;
  imgSrc: string;
  name: string;
  orgName: string;
  description?: string;
  className?: string;
  poapTokenId?: string;
  onClick?: () => void;
};

const LineClamp = (lines: number) => css`
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: ${lines};
  -webkit-box-orient: vertical;
`;

const Wrapper = styled.div`
  display: inline-flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const Info = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  max-width: ${rem(150)};
`;

const TitleStyled = styled(Title)`
  margin-top: ${rem(10)};
  text-align: center;
  ${LineClamp(3)};
`;

const OrgName = styled.div`
  font-family: PT Mono;
  font-style: normal;
  font-weight: bold;
  font-size: ${rem(11)};
  line-height: ${rem(18)};
  text-align: center;
  letter-spacing: ${rem(1.2)};
  text-transform: uppercase;
  color: ${TextLight};
  margin-top: ${rem(8)};
`;

const Description = styled.div`
  font-family: PT Mono;
  font-style: normal;
  font-weight: normal;
  font-size: ${rem(11)};
  line-height: ${rem(14)};
  text-align: center;
  letter-spacing: ${rem(-0.1)};
  color: ${TextLight};
  margin-top: ${rem(8)};
  ${LineClamp(2)};
`;

const Heart = styled(FeatureHeart)`
  position: absolute;
  bottom: ${rem(0)};
  right: ${rem(10)};
`;

const BadgeWrapper = styled(Wrapper)`
  position: relative;
`;

export const GitPOAP = ({
  className,
  poapTokenId,
  gitPOAPId,
  imgSrc,
  name,
  orgName,
  description,
  onClick,
}: Props) => {
  return (
    <Wrapper className={className}>
      <BadgeWrapper>
        <Link href={`/gp/${gitPOAPId}`} passHref>
          <GitPOAPBadge
            size="sm"
            imgUrl={imgSrc}
            onClick={() => {
              if (onClick) {
                onClick();
              }
            }}
          />
          {poapTokenId && <Heart poapTokenId={poapTokenId} />}
        </Link>
      </BadgeWrapper>
      <Info>
        <Link href={`/gp/${gitPOAPId}`} passHref>
          <TitleStyled>{name}</TitleStyled>
        </Link>
        <OrgName>{orgName}</OrgName>
        {description && <Description>{description}</Description>}
      </Info>
    </Wrapper>
  );
};
