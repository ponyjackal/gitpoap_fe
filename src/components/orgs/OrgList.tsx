import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { ItemList, SelectOption } from '../shared/compounds/ItemList';
import { OrgList as OrgListContainer } from '../shared/compounds/OrgList';
import { OrganizationHex, OrganizationHexSkeleton } from './OrgHex';
import {
  OrganizationsListQuery,
  useOrganizationsListQuery,
  useTotalOrganizationCountQuery,
} from '../../graphql/generated-gql';
import { useDebouncedValue } from '@mantine/hooks';
import { Header, Input, TextSkeleton } from '../shared/elements';

type SortOptions = 'alphabetical' | 'date';

const selectOptions: SelectOption<SortOptions>[] = [
  { value: 'alphabetical', label: 'Alphabetical' },
  { value: 'date', label: 'Creation Date' },
];

const StyledHeader = styled(Header)`
  display: block;
  margin-bottom: ${rem(40)};
  max-width: 100%;
`;

const HeaderSkeleton = styled(TextSkeleton)`
  display: block;
  margin-bottom: ${rem(40)};
  width: ${rem(280)};
  max-width: 100%;
`;

const Wrapper = styled.div`
  margin-top: ${rem(80)};
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const StyledItemList = styled(ItemList)`
  margin-bottom: ${rem(50)};
`;

const StyledSearch = styled(Input)`
  margin-bottom: ${rem(40)};
  width: ${rem(400)};
  max-width: 100%;
`;

export type Org = Exclude<OrganizationsListQuery['allOrganizations'], undefined | null>[number];

type QueryVars = {
  sort: SortOptions;
  search: string;
  perPage: number;
  page: number;
};

export const OrgList = () => {
  const [searchValue, setSearchValue] = useState('');
  const [debouncedSearch] = useDebouncedValue(searchValue, 200);

  const [variables, setVariables] = useState<QueryVars>({
    sort: 'date',
    search: '',
    perPage: 15,
    page: 1,
  });
  const [orgListItems, setOrgListItems] = useState<Org[]>([]);
  const [result] = useOrganizationsListQuery({ variables });
  const [totalResult] = useTotalOrganizationCountQuery({
    variables: {
      search: variables.search,
    },
  });
  const allOrganizations = result.data?.allOrganizations;
  const total = totalResult.data?.aggregateOrganization._count?.id;
  const queryVariables = result.operation?.variables;
  const hasMore = !!total && variables.page * variables.perPage < total;

  /* Hook to append new data onto existing list of orgs */
  useEffect(() => {
    const newOrgListItems = allOrganizations ?? [];
    if (queryVariables?.page === 1) {
      setOrgListItems(newOrgListItems);
    } else {
      setOrgListItems((prevOrgListItems) => [...prevOrgListItems, ...newOrgListItems]);
    }
  }, [allOrganizations, queryVariables?.page]);

  useEffect(() => {
    const search = debouncedSearch.length >= 2 ? debouncedSearch : '';
    setVariables((prevVariables) => ({
      ...prevVariables,
      page: 1,
      search,
    }));
  }, [debouncedSearch]);

  if (result.error) {
    return null;
  }

  return (
    <Wrapper>
      {total !== undefined ? (
        <StyledHeader>{`${total} Organizations`}</StyledHeader>
      ) : (
        <HeaderSkeleton height={rem(48)} />
      )}
      <StyledSearch
        placeholder={'SEARCH FOR AN ORGANIZATION...'}
        value={searchValue}
        onChange={(e: React.ChangeEvent<HTMLInputElement>) => setSearchValue(e.target.value)}
      />
      <StyledItemList
        selectOptions={selectOptions}
        selectValue={variables.sort}
        onSelectChange={(sortValue) => {
          if (sortValue !== variables.sort) {
            setVariables({
              ...variables,
              sort: sortValue as SortOptions,
              page: 1,
            });
          }
        }}
        isLoading={result.fetching}
        hasShowMoreButton={hasMore}
        showMoreOnClick={() => {
          if (!result.fetching) {
            setVariables({
              ...variables,
              page: variables.page + 1,
            });
          }
        }}
      >
        <OrgListContainer>
          {result.fetching && !result.operation && (
            <>
              {[...Array(10)].map((_, i) => (
                <OrganizationHexSkeleton key={i} />
              ))}
            </>
          )}

          {orgListItems?.map((org, i) => (
            <OrganizationHex key={'organization-' + i} org={org} />
          ))}
        </OrgListContainer>
      </StyledItemList>
    </Wrapper>
  );
};
