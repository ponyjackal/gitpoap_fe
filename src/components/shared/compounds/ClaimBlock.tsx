import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Button } from '../elements/Button';
import { GitPOAP } from './GitPOAP';
import { FaCheckCircle, FaCoins, FaEthereum } from 'react-icons/fa';

type Props = {
  gitPOAPId: number;
  imgSrc: string;
  name: string;
  orgName: string;
  description: string;
  onClickClaim: () => void;
  onClickBadge?: () => void;
  isClaimed?: boolean;
  isLoading?: boolean;
  isConnected?: boolean;
};

const Wrapper = styled.div`
  display: inline-flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-start;
  height: ${rem(340)};
`;

const ButtonWrapper = styled.div`
  flex: 1;
  margin-top: ${rem(14)};
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
`;

const getButtonText = (isClaimed: boolean | undefined, isConnected: boolean | undefined) => {
  if (isClaimed) {
    return 'Minted';
  }

  if (!isConnected) {
    return 'Connect';
  }

  return 'Mint';
};

export const ClaimBlock = ({
  gitPOAPId,
  imgSrc,
  name,
  orgName,
  description,
  onClickClaim,
  onClickBadge,
  isClaimed,
  isLoading,
  isConnected,
}: Props) => {
  return (
    <Wrapper>
      <GitPOAP
        gitPOAPId={gitPOAPId}
        imgSrc={imgSrc}
        name={name}
        orgName={orgName}
        description={description}
        onClick={onClickBadge}
      />
      <ButtonWrapper>
        <Button
          onClick={onClickClaim}
          loading={isLoading}
          leftIcon={isClaimed ? <FaCheckCircle /> : !isConnected ? <FaEthereum /> : <FaCoins />}
          disabled={isClaimed}
        >
          {getButtonText(isClaimed, isConnected)}
        </Button>
      </ButtonWrapper>
    </Wrapper>
  );
};
