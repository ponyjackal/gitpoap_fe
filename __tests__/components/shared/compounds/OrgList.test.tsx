import { render } from '@testing-library/react';
import 'jest-styled-components';
import { OrgList } from '../../../../src/components/shared/compounds/OrgList';

describe('OrgList', () => {
  it('should render children', () => {
    const { container } = render(
      <OrgList>
        <div>{'test1'}</div>
        <div>{'test2'}</div>
      </OrgList>,
    );
    const orgListContainer = container.firstChild;
    expect(orgListContainer).toBeInTheDocument();
    expect(orgListContainer).toHaveTextContent('test1');
    expect(orgListContainer).toHaveTextContent('test2');
    expect(orgListContainer).toMatchSnapshot();
  });
});
