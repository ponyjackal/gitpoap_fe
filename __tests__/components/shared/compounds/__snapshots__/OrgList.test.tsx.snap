// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrgList should render children 1`] = `
.c0 {
  display: grid;
  width: 100%;
  margin-top: 3.125rem;
  margin-bottom: 3.4375rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  grid-template-columns: repeat(auto-fill,17.5rem);
  -webkit-column-gap: 1.875rem;
  column-gap: 1.875rem;
  row-gap: 2rem;
}

.c1 {
  grid-template-columns: repeat(auto-fill,16.25rem);
}

@media (max-width:50rem) {
  .c0 {
    grid-template-columns: repeat(auto-fit,minmax(15.625rem,1fr));
  }
}

@media (max-width:50rem) {
  .c1 {
    grid-template-columns: unset;
  }
}

<div
  class="c0 c1"
>
  <div>
    test1
  </div>
  <div>
    test2
  </div>
</div>
`;
